vispy/visuals/filters/mesh.py

```diff
@@ -121,104 +124,258 @@ def _attach(self, visual):
     }
-    vec3 specular_color = $light_color * $specular_color * speculark;
-
-    if ($shading_enabled == 1) {
-        vec3 color = $ambient_color + diffuse_color + specular_color;
-        gl_FragColor *= vec4(color, 1.0);
+    normal = normalize(normal);
+
+    vec3 light_vec = normalize(v_light_vec);
+    vec3 eye_vec = normalize(v_eye_vec);
+
+    vec3 ambient = ambient_coeff.rgb * ambient_coeff.a
+                   * ambient_light.rgb * ambient_light.a;
+
+    float diffuse_factor = dot(light_vec, normal);
+    diffuse_factor = max(diffuse_factor, 0.0);
+    vec3 diffuse = diffuse_factor
+                   * diffuse_coeff.rgb * diffuse_coeff.a
+                   * diffuse_light.rgb * diffuse_light.a;
+
+    float specular_factor = 0.0;
+    bool is_illuminated = diffuse_factor > 0.0;
+    if (is_illuminated && shininess > 0.0) {
+        vec3 reflection = reflect(light_vec, normal);
+        specular_factor = dot(reflection, eye_vec);
+        specular_factor = max(specular_factor, 0.0);
+        specular_factor = pow(specular_factor, shininess);
     }
+    vec3 specular = specular_factor
+                    * specular_coeff.rgb * specular_coeff.a
+                    * specular_light.rgb * specular_light.a;
+
+    // XXX(asnt): Not sure if there is a physically more correct way of
+    // blending the base color with the lighting.
+    //vec3 color = base_color * (ambient + diffuse + specular);
+    vec3 color = ambient + base_color * diffuse + specular;
+    gl_FragColor.rgb = color;
 }
 """  # noqa
 
 
+def _as_rgba(intensity_or_color, default_rgb=(1.0, 1.0, 1.0)):
+    """Create an RGBA color from a color or a scalar intensity.
+
+    Examples
+    --------
+    >>> # Specify the full RGBA color.
+    >>> _as_rgba((0.2, 0.3, 0.4, 0.25))
+    ... <Color: (0.2, 0.3, 0.4, 0.25)>
+    >>> # Specify an RGB color. (Default intensity `1.0` is used.)
+    >>> _as_rgba((0.2, 0.3, 0.4))
+    ... <Color: (0.2, 0.3, 0.4, 1.0)>
+    >>> # Specify an intensity only. (Default color `(1.0, 1.0, 1.0)` is used.)
+    >>> _as_rgba(0.25)
+    ... <Color: (1.0, 1.0, 1.0, 0.25)>
+    """
+    if isinstance(intensity_or_color, numbers.Number):
+        intensity = intensity_or_color
+        return Color(default_rgb, alpha=intensity)
+    color = intensity_or_color
+    return Color(color)
+
+
 class ShadingFilter(Filter):
-    """Filter to apply shading to a mesh.
+    """Apply shading to a mesh with the Phong reflection model.
 
     To disable shading, either detach (ex. ``mesh.detach(filter_obj)``) or
     set the shading type to ``None`` (ex. ``filter_obj.shading = None``).
 
+    The shading filter should be attached after the other filters that modify
+    the colors to be shaded. For example, to include the wireframe in the
+    shading, the shading filter must come before the wireframe filter.
+
+    Parameters
+    ----------
+    shading : str
+        Shading mode: None, 'flat' or 'smooth'. If None, the shading is
+        disabled.
```

Comment:
@djhoese I wonder if this parameter/property should not be renamed `mode` (or similar) and take values in `('flat', 'smooth')` and not `None`, which is currently used to disable the filter. Then, we could use an `enabled` property as in `WireframeFilter`. What do you think?

Reply:
I had the impression that `ShadingFilter.shading` was redundant and didn't carry the effect of the parameter clearly but not sure that `ShadingFilter.mode` is better.
Maybe we can keep `shading` as it is and add an `enabled` property anyway for clarity and consistency with `WireframeFilter`? Then the filter would only be applied if `enabled and shading is not None`.

Reply:
Oh naming. What a fun set of decisions to make.

@kmuehlbauer @larsoner do you have any feelings about these names? I agree consistency between the different mesh filters to use `.enabled` would be nice. Regardless of how the filter is done though we probably want backwards compatibility with the MeshVisual's shading. So if the filter is different the property on the MeshVisual still needs to map to the same options to not break compatibility. `.mode` or `.method` or something else is fine by me.

Reply:
I think keeping it as `shading` is for the best. At least for now. Looking at the MeshVisual, it already has a `mode` property which may get confusing. And I think shading should keep all three options for now. In the future if we want to remove shading from the MeshVisual and make users have to add the filter for any shading then we can change what the shading parameter accepts.